File: bids_exporter/src/infrastructure/object_storage.py
import asyncio
from minio import Minio as S3CompatibleClient

from ..config.env import settings


object_storage_client = S3CompatibleClient(
    endpoint=f"{settings.object_storage_endpoint}:{settings.object_storage_port}",
    access_key=settings.object_storage_access_key,
    secret_key=settings.object_storage_secret_key,
    secure=settings.object_storage_use_ssl,
)

RAW_DATA_BUCKET = settings.object_storage_raw_data_bucket
MEDIA_BUCKET = settings.object_storage_media_bucket
BIDS_BUCKET = settings.object_storage_bids_exports_bucket


async def check_object_storage_connection() -> None:
    """Ensure the object storage connection succeeds and the BIDS bucket exists."""
    print("Checking object storage connection and bucket status...")
    loop = asyncio.get_event_loop()
    try:
        found = await loop.run_in_executor(None, object_storage_client.bucket_exists, BIDS_BUCKET)
        if not found:
            print(f"Bucket '{BIDS_BUCKET}' not found. Creating it...")
            await loop.run_in_executor(None, object_storage_client.make_bucket, BIDS_BUCKET)
            print(f"Bucket '{BIDS_BUCKET}' created successfully.")
        else:
            print(f"Bucket '{BIDS_BUCKET}' already exists.")

        print("✅ Object storage connection is OK.")

    except Exception as exc:  # noqa: BLE001
        print(f"❌ FATAL: Could not connect to object storage or set up the bucket: {exc}")
        raise

File: stimulus_asset_processor/src/infrastructure/objectStorage.ts
import { Client as S3CompatibleClient } from 'minio'

import { config } from '../config/env'
import {
  describeStorageError,
  ensureBucketExists,
  waitForObjectStorageConnection,
} from '../../../shared/objectStorage'

export const objectStorageClient = new S3CompatibleClient({
  endPoint: config.OBJECT_STORAGE_ENDPOINT,
  port: config.OBJECT_STORAGE_PORT,
  useSSL: config.OBJECT_STORAGE_USE_SSL,
  accessKey: config.OBJECT_STORAGE_ACCESS_KEY,
  secretKey: config.OBJECT_STORAGE_SECRET_KEY,
})

/**
 * サービス起動時に、必要なオブジェクトストレージのバケットが存在することを保証します。
 * 存在しない場合は作成を試みます。
 */
export async function ensureObjectStorageBucket(
  maxAttempts = 5,
  baseDelayMs = 1_000,
): Promise<void> {
  const connectionAttempts = Math.max(maxAttempts, 10)

  await waitForObjectStorageConnection(objectStorageClient, {
    maxAttempts: connectionAttempts,
    baseDelayMs,
    onRetry: ({ attempt, maxAttempts: max, error }) =>
      console.warn(
        `⏳ [ObjectStorage] Waiting for endpoint (attempt ${attempt}/${max}). reason=${describeStorageError(error)}`,
      ),
    onSuccess: ({ attempt }) => {
      if (attempt > 1) {
        console.log(`✅ [ObjectStorage] Connected on attempt ${attempt}.`)
      }
    },
    onFailure: ({ error }) =>
      console.error('❌ [ObjectStorage] Failed to connect to object storage.', error),
  })

  await ensureBucketExists(objectStorageClient, config.OBJECT_STORAGE_MEDIA_BUCKET, {
    maxAttempts,
    baseDelayMs,
    onCreateStart: () =>
      console.warn(
        `[ObjectStorage] Bucket "${config.OBJECT_STORAGE_MEDIA_BUCKET}" does not exist. Creating...`,
      ),
    onCreateSuccess: () =>
      console.log(`✅ [ObjectStorage] Bucket "${config.OBJECT_STORAGE_MEDIA_BUCKET}" created successfully.`),
    onAlreadyExists: () =>
      console.log(`✅ [ObjectStorage] Bucket "${config.OBJECT_STORAGE_MEDIA_BUCKET}" is ready.`),
    onRetry: ({ attempt, maxAttempts: max, error }) =>
      console.warn(
        `⏳ [ObjectStorage] Bucket check will retry (attempt ${attempt}/${max}). reason=${describeStorageError(error)}`,
      ),
    onFailure: ({ attempt, maxAttempts: max, error }) =>
      console.error(
        `❌ [ObjectStorage] Failed to ensure bucket exists (attempt ${attempt}/${max}).`,
        error,
      ),
  })
}
